File: IsometricRPG/IsometricRPG/Utils/IsometricMath.swift
import CoreGraphics

/// Utility for converting between Cartesian (game world) and isometric (screen) coordinates.
enum IsometricMath {
    /// Convert a grid position (col, row) to screen-space isometric coordinates.
    static func gridToScreen(col: Int, row: Int) -> CGPoint {
        let x = CGFloat(col - row) * (Constants.tileWidth / 2)
        let y = CGFloat(col + row) * (Constants.tileHeight / 2)
        return CGPoint(x: x, y: -y) // SpriteKit y-up, iso y-down
    }

    /// Convert a world position (continuous) to screen-space isometric coordinates.
    static func worldToScreen(_ point: CGPoint) -> CGPoint {
        let x = (point.x - point.y) * (Constants.tileWidth / 2)
        let y = (point.x + point.y) * (Constants.tileHeight / 2)
        return CGPoint(x: x, y: -y)
    }

    /// Convert screen-space isometric coordinates back to world position.
    static func screenToWorld(_ point: CGPoint) -> CGPoint {
        let adjustedY = -point.y
        let x = (point.x / (Constants.tileWidth / 2) + adjustedY / (Constants.tileHeight / 2)) / 2
        let y = (adjustedY / (Constants.tileHeight / 2) - point.x / (Constants.tileWidth / 2)) / 2
        return CGPoint(x: x, y: y)
    }

    /// Distance between two CGPoints.
    static func distance(_ a: CGPoint, _ b: CGPoint) -> CGFloat {
        let dx = b.x - a.x
        let dy = b.y - a.y
        return sqrt(dx * dx + dy * dy)
    }

    /// Normalized direction vector from `from` to `to`.
    static func direction(from: CGPoint, to: CGPoint) -> CGPoint {
        let dx = to.x - from.x
        let dy = to.y - from.y
        let len = sqrt(dx * dx + dy * dy)
        guard len > 0 else { return .zero }
        return CGPoint(x: dx / len, y: dy / len)
    }
}

File: IsometricRPGMono/Utils/IsometricMath.cs
using Microsoft.Xna.Framework;

namespace IsometricRPG;

public static class IsometricMath
{
    public static Vector2 GridToScreen(int col, int row, int elevation = 0)
    {
        float x = (float)(col - row) * (Constants.TileWidth / 2f);
        float baseY = (float)(col + row) * (Constants.TileHeight / 2f);
        float heightOffset = (float)elevation * Constants.ElevationHeightMultiplier;
        float y = baseY - heightOffset;
        // MonoGame is Y-down: no negation of y
        return new Vector2(x, y);
    }

    public static Vector2 WorldToScreen(Vector2 point)
    {
        float x = (point.X - point.Y) * (Constants.TileWidth / 2f);
        float y = (point.X + point.Y) * (Constants.TileHeight / 2f);
        // MonoGame is Y-down: no negation of y
        return new Vector2(x, y);
    }

    public static Vector2 ScreenToWorld(Vector2 point)
    {
        // MonoGame Y-down: use point.Y directly (no negation)
        float x = (point.X / (Constants.TileWidth / 2f) + point.Y / (Constants.TileHeight / 2f)) / 2f;
        float y = (point.Y / (Constants.TileHeight / 2f) - point.X / (Constants.TileWidth / 2f)) / 2f;
        return new Vector2(x, y);
    }

    public static float Distance(Vector2 a, Vector2 b)
    {
        float dx = b.X - a.X;
        float dy = b.Y - a.Y;
        return MathF.Sqrt(dx * dx + dy * dy);
    }

    public static Vector2 Direction(Vector2 from, Vector2 to)
    {
        float dx = to.X - from.X;
        float dy = to.Y - from.Y;
        float len = MathF.Sqrt(dx * dx + dy * dy);
        if (len < 1e-6f) return Vector2.Zero;
        return new Vector2(dx / len, dy / len);
    }
}
